crates/net/network/src/transactions/mod.rs

```diff
@@ -1082,84 +1082,113 @@ where
 
-            for res in import_res {
-                match res {
-                    Ok(hash) => {
-                        this.on_good_import(hash);
-                    }
+            // Advance all imports
+            if let Poll::Ready(Some(import_res)) = this.pool_imports.poll_next_unpin(cx) {
+                let import_res = match import_res {
+                    Ok(res) => res,
                     Err(err) => {
-                        // if we're _currently_ syncing and the transaction is bad we ignore it,
-                        // otherwise we penalize the peer that sent the bad
-                        // transaction with the assumption that the peer should have
-                        // known that this transaction is bad. (e.g. consensus
-                        // rules)
-                        if err.is_bad_transaction() && !this.network.is_syncing() {
-                            debug!(target: "net::tx", ?err, "bad pool transaction import");
-                            this.on_bad_import(err.hash);
-                            continue
+                        debug!(target: "net::tx", ?err, "bad pool transaction batch import");
+                        continue
+                    }
+                };
+
+                for res in import_res {
+                    match res {
+                        Ok(hash) => {
+                            this.on_good_import(hash);
+                        }
+                        Err(err) => {
+                            // if we're _currently_ syncing and the transaction is bad we ignore it,
+                            // otherwise we penalize the peer that sent the bad
+                            // transaction with the assumption that the peer should have
+                            // known that this transaction is bad. (e.g. consensus
+                            // rules)
+                            if err.is_bad_transaction() && !this.network.is_syncing() {
+                                debug!(target: "net::tx", ?err, "bad pool transaction import");
+                                this.on_bad_import(err.hash);
+                                continue
+                            }
+                            this.on_good_import(err.hash);
                         }
-                        this.on_good_import(err.hash);
                     }
                 }
+                some_ready = true;
             }
-        }
 
-        // handle and propagate new transactions
-        let mut new_txs = Vec::new();
-        while let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
-            new_txs.push(hash);
-        }
-        if !new_txs.is_empty() {
-            this.on_new_transactions(new_txs);
-        }
+            // handle and propagate new transactions
+            let mut new_txs = Vec::new();
+            if let Poll::Ready(Some(hash)) = this.pending_transactions.poll_next_unpin(cx) {
```

Comment:
this should still drain the entire buffer so we can send as many as possible

Reply:
how about adding an internal budget for that one then. from a DoS perspective, if there are endless txns to propagate, this future is blocking if we drain the whole thing. this could also happen to some degree if the rest of the node's work isn't perfectly synchronised. so if adding txns to pending transactions is too efficient with respect to the rest of processing.

Reply:
it makes sense to tune this a bit and not poll this on every invocation.
we could/should also use the channel that streams full transactions so we don't have to lookup the txs again

but for now I want to drain the receiver buffer here